### PR TITLE
segments floats or timedelta

### LINKED ISSUE #133: segment function on time objects
The `segment` function is not able to handle `datetime` objects. This needs to work, as an example for the GDP dataset. The following should work, but does not:

```
import xarray as xr
from clouddrift.analysis import segment
from datetime import timedelta
ds = xr.open_dataset("gdp_v2.00.nc")
traj_idx = np.insert(np.cumsum(ds.rowsize.values), 0, 0)
j = 100
sli = slice(traj_idx[j], traj_idx[j+1])
dt = timedelta(seconds=1800)
row1 = segment(time[sli],tolerance=dt)
```

### PR DESCRIPTION
**Pull request by philippemiron**
Fix #133

### CODE REVIEW

File: tests/analysis_tests.py [hunk 1]
```diff
@@ -45,6 +45,17 @@ def test_segment_rowsize_raises(self):
         with self.assertRaises(ValueError):
             segment(x, tol, rowsize)
 
+    def test_segments_datetime(self):
+        x = [
+            np.datetime64("2023-01-01"),
+            np.datetime64("2023-01-02"),
+            np.datetime64("2023-01-03"),
+            np.datetime64("2023-02-01"),
+            np.datetime64("2010-02-01"),
```
Comment (selipot) [inline tests/analysis_tests.py:54]: I think there's an issue here. Did you mean "2023-02-02"?

Comment (selipot) [reply tests/analysis_tests.py:54]: Let's put a test with hour intervals. We want the tolerance to be the maximum difference allowed so that `tol = np.timedelta64(1,"H")` will split the GDP hourly ragged array, correct?

File: tests/analysis_tests.py [hunk 2]
```diff
@@ -45,6 +45,17 @@ def test_segment_rowsize_raises(self):
         with self.assertRaises(ValueError):
             segment(x, tol, rowsize)
 
+    def test_segments_datetime(self):
+        x = [
+            np.datetime64("2023-01-01"),
```
Comment (selipot) [inline tests/analysis_tests.py:50]: the function uses `datetime.timedelta` but here you use `np.datetime64`. Does it matter?

**Review by selipot — CHANGES_REQUESTED** [review]
I will test soon with the GDP!

File: tests/analysis_tests.py (hunk 1 above)
Comment (philippemiron) [reply tests/analysis_tests.py:54]: Yes, that was a typo. I'm not sure I understand the difference with the test you want to do (apart from 1h instead of 1d).

File: tests/analysis_tests.py (hunk 2 above)
Comment (philippemiron) [reply tests/analysis_tests.py:50]: I modified to datetime for consistency, but looks like both are compatible.

File: clouddrift/analysis.py [hunk 3]
```diff
@@ -91,7 +92,9 @@ def apply_ragged(
 
 
 def segment(
-    x: np.ndarray, tolerance: float, rowsize: np.ndarray[int] = None
+    x: np.ndarray,
+    tolerance: float | timedelta,
```
Comment (milancurcic) [inline clouddrift/analysis.py:96]: This syntax was introduced in Python 3.10. If we want to support Python 3.9, use Union:

```suggestion
    tolerance: Union[float, timedelta],
```

(but I personally am fine dropping 3.9 for improved type hint syntax)

File: clouddrift/analysis.py [hunk 4]
```diff
@@ -147,8 +150,13 @@ def segment(
     array([2, 2, 2, 2])
     """
 
+    if type(tolerance) == timedelta:
+        positive_tol = tolerance > timedelta(seconds=0)
+    else:
+        positive_tol = tolerance > 0
```
Comment (milancurcic) [inline clouddrift/analysis.py:156]: Don't you need >= here? Otherwise `tolerance=0` would produce incorrect results, it seems.

```suggestion
    if type(tolerance) == timedelta:
        positive_tol = tolerance >= timedelta(seconds=0)
    else:
        positive_tol = tolerance >= 0
```

**Review by milancurcic — APPROVED** [review]
Looks good pending the review getting resolved.

File: clouddrift/analysis.py (hunk 4 above)
Comment (milancurcic) [reply clouddrift/analysis.py:156]: Let's also add a unit test for tolerance of positive and negative zero.

Comment (philippemiron) [reply clouddrift/analysis.py:156]: Positive or negative 0?

Comment (philippemiron) [reply clouddrift/analysis.py:156]: If tolerance is 0, it should just return `np.arange(x)` no? Maybe it should be an ArgumentError.

File: clouddrift/analysis.py (hunk 3 above)
Comment (philippemiron) [reply clouddrift/analysis.py:96]: I will make the tests run with 3.9 and 3.10 for now. So we can catch this type of error.. because I'm only using 3.10 (or 3.11) currently.

File: clouddrift/analysis.py (hunk 4 above)
Comment (milancurcic) [reply clouddrift/analysis.py:156]: Please ignore my mention of a negative zero. I did some reading and see that both the IEEE standard and Python define signed zero such that `-0 == 0`.

If tolerance is 0, we tolerate consecutive values equal to each other to be a segment. So, `segment([1, 1, 2, 2], tolerance=0)` should return `[2, 2]`.

Comment (selipot) [reply clouddrift/analysis.py:156]: That's a good example right there to form a rowsize variable if it does not exist.

**Comment by philippemiron** [comment]
I'm not sure why, but I marked conversations as resolved but the PR is still blocked.

**Comment by selipot** [comment]
> I'm not sure why, but I marked conversations as resolved but the PR is still blocked.

Maybe because I requested some changes?

**Review by selipot — APPROVED** [review]

**Comment by philippemiron** [comment]
I thought the comments being resolved would change to approve, but it doesn't.
